modules/core/src/main/java/org/apache/ignite/internal/client/thin/TcpClientCache.java

```diff
@@ -1616,4 +1650,77 @@ private void checkDataReplicationSupported(ProtocolContext protocolCtx)
         if (!protocolCtx.isFeatureSupported(ProtocolBitmaskFeature.DATA_REPLICATION_OPERATIONS))
             throw new ClientFeatureNotSupportedByServerException(ProtocolBitmaskFeature.DATA_REPLICATION_OPERATIONS);
     }
+
+    /**
+     * Warns if an unordered map is used in an operation that may lead to a distributed deadlock
+     * during an explicit transaction.
+     * <p>
+     * This check is relevant only for explicit user-managed transactions. Implicit transactions
+     * (such as those started automatically by the system) are not inspected by this method.
+     * </p>
+     *
+     * @param m        The map being used in the cache operation.
+     */
+    protected void warnIfUnordered(Map<?, ?> m) {
+        if (m == null || m.size() <= 1)
+            return;
+
+        TcpClientTransaction tx = transactions.tx();
+
+        // Only explicit transactions are checked
+        if (tx == null)
+            return;
+
+        if (m instanceof SortedMap || m instanceof GridSerializableMap)
```

Comment:
GridSerializableMap - not sure about this class, looks like required only on server side